File: vbnet/metafiles_gui/metafiles_gui.vb
' metafiles_gui -- VB.NET port of examples\Vb6\metafiles_gui (headless Sub Main)
Imports System
Imports LumasPdfSdk

Module modMetafilesGui
    Private errDel As TErrorProc
    Private Const MARGIN As Double = 10.0

    Function AppPath() As String
        Return AppDomain.CurrentDomain.BaseDirectory.TrimEnd("\"c)
    End Function

    Public Function ErrProc(ByVal Data As IntPtr, ByVal ErrCode As Integer, ByVal ErrMessage As String, ByVal ErrType As Integer) As Integer
        Console.WriteLine(ErrMessage)
        Return 0
    End Function

    Private Sub PlaceEMFCentered(ByVal pdf As IntPtr, ByVal mFile As String, ByVal Width_ As Double, ByVal Height_ As Double)
        Dim x, y, w, h, sx As Double
        Dim r As TRectL
        LumasPdf.pdfGetLogMetafileSize(pdf, mFile, r)
        w = r.Right - r.Left
        h = r.Bottom - r.Top
        Width_ = Width_ - 2.0 * MARGIN
        Height_ = Height_ - 2.0 * MARGIN
        sx = Width_ / w
        If (h * sx <= Height_) Then
            x = MARGIN
            y = MARGIN
            LumasPdf.pdfInsertMetafile(pdf, mFile, x, y, Width_, 0.0)
        Else
            sx = Height_ / h
            w = w * sx
            x = MARGIN + (Width_ - w) / 2.0
            y = MARGIN
            LumasPdf.pdfInsertMetafile(pdf, mFile, x, y, 0.0, Height_)
        End If
    End Sub

    Sub Main()
        Dim pdf As IntPtr = LumasPdf.pdfNewPDF()
        errDel = New TErrorProc(AddressOf ErrProc)
        LumasPdf.pdfSetOnErrorProc(pdf, IntPtr.Zero, errDel)
        LumasPdf.pdfSetCompressionFilter(pdf, LumasPdfConsts.cfFlate)
        LumasPdf.pdfSetJPEGQuality(pdf, 70)

        Dim inFile As String = AppPath() & "\in.emf"
        Dim outFile As String = AppPath() & "\out.pdf"

        If LumasPdf.pdfCreateNewPDFW(pdf, "") = 0 Then
            LumasPdf.pdfDeletePDF(pdf)
            Return
        End If

        LumasPdf.pdfSetCompressionLevel(pdf, TCompressionLevel.clNone)
        LumasPdf.pdfSetCompressionFilter(pdf, LumasPdfConsts.cfFlate)
        LumasPdf.pdfSetColorSpace(pdf, TPDFColorSpace.csDeviceRGB)
        LumasPdf.pdfSetMetaConvFlags(pdf, LumasPdfConsts.mfDefault)
        LumasPdf.pdfSetPageCoords(pdf, TPageCoord.pcTopDown)
        LumasPdf.pdfAppend(pdf)
        LumasPdf.pdfSetResolution(pdf, 300)
        LumasPdf.pdfSetJPEGQuality(pdf, 70)
        PlaceEMFCentered(pdf, inFile, LumasPdf.pdfGetPageWidth(pdf), LumasPdf.pdfGetPageHeight(pdf))
        LumasPdf.pdfEndPage(pdf)

        If LumasPdf.pdfHaveOpenDoc(pdf) <> 0 Then
            If LumasPdf.pdfOpenOutputFileW(pdf, outFile) = 0 Then
                LumasPdf.pdfFreePDF(pdf)
                LumasPdf.pdfDeletePDF(pdf)
                Return
            End If
            If LumasPdf.pdfCloseFile(pdf) <> 0 Then
                Console.WriteLine("OK: " & outFile)
            End If
        End If

        LumasPdf.pdfDeletePDF(pdf)
    End Sub
End Module

File: cpp-linux/metafiles_gui/metafiles_gui.cpp
// metafiles_gui -- C++ port of examples\Vb6\metafiles_gui\metafiles_gui.bas
// The Delphi original was a GUI metafile viewer/converter; here the core logic is a
// plain main(): load an EMF, place it centered on a page, write out.pdf.
#include <lumaspdf.h>
#include "repo_root.h"
#include <cstdio>
#include <string>

static const double MARGIN = 10.0;

static std::string exeDir(const char* a0){ std::string s(a0); auto p=s.find_last_of("\\/"); return p==std::string::npos?".":s.substr(0,p); }

static SI32 PDF_CALL ErrProc(void* Data, SI32 ErrCode, const char* ErrMessage, SI32 ErrType){
    if(ErrMessage) printf("%s\n", ErrMessage);
    return 0;
}

static void PlaceEMFCentered(PPDF pdf, const char* mFile, double width, double height){
    TRectL r{};
    pdfGetLogMetafileSize(pdf, mFile, &r);
    double w = r.Right - r.Left;
    double h = r.Bottom - r.Top;
    width  -= 2.0 * MARGIN;
    height -= 2.0 * MARGIN;
    double sx = width / w;
    if(h * sx <= height){
        double x = MARGIN, y = MARGIN;
        pdfInsertMetafile(pdf, mFile, x, y, width, 0.0);
    } else {
        sx = height / h;
        w = w * sx;
        double x = MARGIN + (width - w) / 2.0;
        double y = MARGIN;
        pdfInsertMetafile(pdf, mFile, x, y, 0.0, height);
    }
}

int main(int argc, char** argv){
    PPDF pdf = pdfNewPDF();
    pdfSetOnErrorProc(pdf, nullptr, ErrProc);
    pdfSetCompressionFilter(pdf, cfFlate);
    pdfSetJPEGQuality(pdf, 70);

    // Original input came from an interactive shell tree; here a fixed file.
    const char* inFile = LUMAS_REPO_ROOT "/sample_vector.emf";
    std::string outFile = exeDir(argv[0]) + "/out.pdf";

    if(pdfCreateNewPDFA(pdf, "") == 0){ pdfDeletePDF(pdf); return 1; }

    pdfSetCompressionLevel(pdf, clNone);
    pdfSetCompressionFilter(pdf, cfFlate);
    pdfSetColorSpace(pdf, csDeviceRGB);
    pdfSetMetaConvFlags(pdf, mfDefault);
    pdfSetPageCoords(pdf, pcTopDown);
    pdfAppend(pdf);
    pdfSetResolution(pdf, 300);
    pdfSetJPEGQuality(pdf, 70);
    PlaceEMFCentered(pdf, inFile, pdfGetPageWidth(pdf), pdfGetPageHeight(pdf));
    pdfEndPage(pdf);

    if(pdfHaveOpenDoc(pdf) != 0){
        if(pdfOpenOutputFileA(pdf, outFile.c_str()) == 0){ pdfFreePDF(pdf); pdfDeletePDF(pdf); return 1; }
        if(pdfCloseFile(pdf) != 0) printf("OK: %s\n", outFile.c_str());
    }
    pdfDeletePDF(pdf);
    return 0;
}
